### PR TITLE
No-trigger mode

### PR DESCRIPTION
**Pull request by diegoasua**

### CODE REVIEW

**Comment by diegoasua** [comment]
While working on this I noticed it is more difficult than what I initially thought. For instance, the `ExternalComm.duration_queue` is pipped directly from the `ExternalComm` to the `Saver` in the initialization of the later process. Therefore such queue has to exist, and the `Mock` object does not have it. Second, and more importantly, how do we deal with settings this _from the GUI_? If this is dealt like in the other modes, one could initialise sashimi with an option for triggerless. In my humble opinion that is not the way to go... Rather, as it is drafted here, it is in the GUI where one specifies whether the triggering will take place or not. Then, which is the way to go? The `ExternalComm` is already initialised and eventhough one could kill the process and redefine and start a new process of either a real or a mock one, that would be tedious and inefficient and lead to some weird complications. Not the way to go either. So, how do we deal with this?

**Comment by vigji** [comment]
I would leave this to the enlightened opinion of @vilim, with no rush on him. I guess we could share the queue also with the interface, but then we can't just rely on when the parameter is changed to convey it to the Saver process, or this can easily lead to bugs when switching from the "external" to the "internal" setting of the duration

**Comment by vilim** [comment]
The duration queue should be initiated in the state (not in the ExternalComm), The saver just uses the last duration put into the queue, so when you disable the external duration receipt with the GUI, you can just keep putting the new, gui-controlled durations on it.

**Review by vilim — CHANGES_REQUESTED** [review]
To me this seems almost done except for the comment I left. In any case, it can all be tested without the light sheet with running stytra on the same computer (ZMQ has no issues with that). Please do this test before merge (recording will probably seem a bit off because of the imperfect simulation of the lightsheet, but it's a good point to test it here anyway).

File: sashimi/state.py
Comment (diegoasua) [reply sashimi/state.py:667]: The flow is as follows. If the widget is checked in the GUI then everything proceeds like before. When unchecked, two things happen:

1. The `State` sets a event in the `ExternalComm` process indicating it to not put the duration on the queue (which happens only if the experiment is set to start)
2. Duration widget is set to `enabled` &mdash; the change in `lightparam.Param` is connected to the function you point to but it does not really send anything while the box is checked because the widget is disabled.

Does this answer your question?

**Comment by diegoasua** [comment]
It gets stuck when triggering the experiment with `Triggered experiment` box unchecked after pressing play. It does not freeze it just stays waiting (for...?) and after some seconds one of the queues gets full and sashimi crashes. **Can someone reproduce this please?** I want to test if this only happens to me, whereby we may conclude it is due to low computational resources (my laptop is not precisely NASA-grade) or if this is (as I suspect) from the software.

If it is coming from sashimi, it might be one of:

1. `Mock Camera`. To test this hypothesis, please @vigji run this branch https://github.com/portugueslab/sashimi/tree/no-trigger-no-life in the actual data acquistion machine and let us know what the Caesar thinks of it, does it 👍 or shall it 👎 to the lions.
2. Inter-process communication. It is increasingly hard to keep track of Queues and Events floating around, a mess wherein one easily could have forgotten to send a variable to all processes that require.
3. External trigger process, but I see no conflicts.
4. As I said, low computational resources in my laptop.

**Comment by vigji** [comment]
I would have really strong doubts on 4.; also 1. is highly suspicious. It really seems to be something that comes from some miscommunication of some funny bug somewhere, I would investigate more thoroughly around those issues. At least to know exactly in which point of the execution the program is stuck, after which more circumstantiated hps can be done on points 1 and 4.

**Comment by diegoasua** [comment]
> I would have really strong doubts on 4.; also 1. is highly suspicious. It really seems to be something that comes from some miscommunication of some funny bug somewhere, I would investigate more thoroughly around those issues. At least to know exactly in which point of the execution the program is stuck, after which more circumstantiated hps can be done on points 1 and 4.

Yes ok but can you reproduce this behaviour to confirm please

**Comment by vigji** [comment]
Yes, I can reproduce on my laptop, with good performance. If I make sure the "Triggered experiment" box is checked, the acquisition actually starts, suggesting that it is unlikely to be a mock camera-related issue.

**Comment by diegoasua** [comment]
> Yes, I can reproduce on my laptop, with good performance. If I make sure the "Triggered experiment" box is checked, the acquisition actually starts, suggesting that it is unlikely to be a mock camera-related issue.

Great thanks. And what happens if you uncheck?

**Comment by vigji** [comment]
It hangs, and crashes because of full queue

**Comment by diegoasua** [comment]
Can someone draft me all the steps that happen since one presses start button to actually running the experiment? I am terribly sorry guys but I am unable to find where this bug lays and I believe it might be because I don't fully understand all the init process. I am trying to trace down the bug but my abilities seem not to be enough for such quest. Ideas welcome. Otherwise I will leave it here as it is for someone else to debug.

**Comment by vilim** [comment]
This is the main reason I built the multiprocessing log system, see this [gist](https://gist.github.com/vilim/10dabc94f836c0d7d00fa25fe90f6303) for how to inspect them

**Comment by vilim** [comment]
Will have a look this evening, we anyway need to have an annotated diagram of it for the docs

**Comment by diegoasua** [comment]
> This is the main reason I built the multiprocessing log system, see this [gist](https://gist.github.com/vilim/10dabc94f836c0d7d00fa25fe90f6303) for how to inspect them

Thank you! Yes I had forgotten this exists. This will be very helpful. I will give another try later today while looking at the log. Thanks Vilim

**Comment by vigji** [comment]
Any update on this Diego? Tomorrow someone (likely, someone with me) will be put to work on the planar mode, so I guess that helping you figuring this out should be the first aim

**Comment by diegoasua** [comment]
> Any update on this Diego? Tomorrow someone (likely, someone with me) will be put to work on the planar mode, so I guess that helping you figuring this out should be the first aim

Yes indeed. The comment before is my update but no one replied

**Comment by vigji** [comment]
With the mock camera there might be something problematic in the conditional statements on camera on/off signals in `scanloops.py`, so that the `wait_signal.clear()` call is never reached. Sorry when I ruled out a `MockCamera` related problem, I was not aware of those checks in `scanloops.py`.
Can you elaborate on the logic of these statements @vilim or point me to somewhere where they were discussed?

**Comment by vigji** [comment]
Sorry for the confusion, I thought those parameters pertained to the camera, but they actually are set only in the scanloop itself. They should probably be renamed?

**Comment by vigji** [comment]
Still looking into what is happening, but for some reason in the MockScanning mode line 381 of the scanloop that clears the wait_signal is never reached

**Comment by vilim** [comment]
Ah I think I know where the issue might be. the actual implementation of the temporal sequence is quite a bit dependent on the real world: the real camera does not output any frames during the wait phase (because the trigger signal is not set), while the mock camera outputs frames continuously, as we are not simulating the external trigger (see comments in the camera process in this branch). It feels like that there might need to be an extra process "mock external world" to test all these things properly. This suggestion is only half in jest.
@diegoasua if the `is_triggered_event` is not enabled, the `experiment_trigger_event` does not get cleared appropriately.
 In general, we should do a sweep through the processes and give the events the same, descriptive names (also coordinated with the SashimiEvents enum). For each event there should be a description when it is supposed to be on, off, and who manages the transitions and on what conditions.

**Comment by vigji** [comment]
But I still don't understand why we are stuck in the `scanloop.loop` after starting the experiment, there should nothing dependent on real/mock camera there no?

**Comment by diegoasua** [comment]
There are two issues here @vigji, that @vilim correctly points. One is the fact that the mock camera has a bit of free spirit in comparison to the real camera. That causes the `ArrayQueue` to crash. The other issue is that `is_triggered_event` blocks the start of the experiment. I don't know the specifics but I suspect `scanloop.loop` initiation depends (indirectly) on `is_triggered_event.set()`.

**Comment by vigji** [comment]
I figured out the key problem, although I still lack a full explanation for what is happening. Importantly though, I have a solution :D

For some reason, having a `MockScanning` board with an instantaneous `read()` method is blocking the while True `ScanLoop.loop` in the after a certain number of repetitions. It seems that just adding a` sleep(0.05)` or something like that in the read of the `MockBoard` fixes it.

**Comment by vigji** [comment]
Why the progress bar got removed @diegoasua ?

**Comment by diegoasua** [comment]
It is the second time we have to introduce sleep times to combat experiment initialization like in

https://github.com/portugueslab/sashimi/blob/df83b9974b78b766cba972f5e0df7387cb2744c6/sashimi/state.py#L564

Something happened with the progress bar when I moved it to the toolbar. Apologies for that. I have some code written from when I was trying to fix experiment start that will probably bring it back, along with some improvements to the toolbar. Will push it later

**Comment by vigji** [comment]
Can you push it now? We wanted to try merging this today to keep going on other things

**Comment by vilim** [comment]
@vigji that _is_ the correct solution, because the NI board pauses the program execution until the samples are available (the actual sleep time is `t_sleep = n_samples/sample_rate - t_processing (from the rest of the loop)`, but getting it to that actual time is probably not necessary. The accumulation of the frames from the mock camera shouldn't be a big issue because even during the wait period the frames are supposed to be dropped, perhaps the size is still too big for testing.
Btw the toolbar should be redisigned, so the 4 modes are in it, not as tabs in the widget below, and selecting a mode reconfigures other widgets and the toolbar itself (so that e.g. the play button only appears in the recording modes)
image

**Comment by vigji** [comment]
Ok, that is great to hear!
What are we missing for this to be good to go apart from restoring the progress bar and ensure the correct duration is set? (at the moment duration does not change when changing it in the tool)

**Comment by vigji** [comment]
For the bar problem. Apparently, the lines 43-44 of top_bar.py:
```
self.experiment_progress.show()
self.lbl_experiment_progress.show()
```
get called but they do not result in the expected behavior (the bar showing). It might be a GUI-refreshing related problem? I would just remove this hiding of the progress bar, and have it constantly there when we are in a mode that allows to run experiments. Is this fine @vilim ? We can figure this out later in a GUI prettification PR

**Comment by kkoetter** [comment]
i committed a couple of changes. the progress bar and experiment start button are now hidden upon paused and preview and appear all the time during both experiment states like @vigji wrote. there are still some bugs when changing tabs, which will have to be figured out. 
Then I also cleared the self.wait_signal in the planar scanning loop and it seems to do the trick for it to not hang anymore.

**Comment by diegoasua** [comment]
> @vigji that _is_ the correct solution, because the NI board pauses the program execution until the samples are available (the actual sleep time is `t_sleep = n_samples/sample_rate - t_processing (from the rest of the loop)`, but getting it to that actual time is probably not necessary. The accumulation of the frames from the mock camera shouldn't be a big issue because even during the wait period the frames are supposed to be dropped, perhaps the size is still too big for testing.
> Btw the toolbar should be redisigned, so the 4 modes are in it, not as tabs in the widget below, and selecting a mode reconfigures other widgets and the toolbar itself (so that e.g. the play button only appears in the recording modes)
> image

This is implemented, will push it later. Unfortunately a 6 h time difference means I was sleeping

**Comment by kkoetter** [comment]
@diegoasua for the manual duration update: it seems to me that the new duration is send to trigger settings and the general distribution queue, but that` n_volumes `in `SavingParameters` is never changed and therefore the status bar doesn't update according to experiment duration. I checked that in `receive_save_parameters` the value actually changes according to the manual input.

**Comment by diegoasua** [comment]
> @diegoasua for the manual duration update: it seems to me that the new duration is send to trigger settings and the general distribution queue, but that`n_volumes`in `SavingParameters` is never changed and therefore the status bar doesn't update according to experiment duration. I checked that in `receive_save_parameters` the value actually changes according to the manual input.
> i think the bug might be in line 97 `self.i_volume < self.save_parameters.n_volumes` and my best guess is the update of manual input is not active for the first round of running, therefore 1000 default is put in place and updated later without effect.

Indeed there is a little bug here. This may happen because even after experiment finishes the MockCamera still records new frames. The update in the progressbar is as you say @kkoetter

**Comment by vigji** [comment]
The problem with the duration is the following:
At the moment we are setting a default panel of `SavingParameters`. The State dispatches them to the `StackSaver` via a new `SavingParameters()` object generated by the call to `convert_save_params()` function in `State`, line 599.
This always dispatches to the `StackSaver` the standard duration, which has always been overwritten by the value in the `duration_queue` which has always been arrived directly  from the `ExternalComm` process. 
Now we are changing it in the GUI before starting the experiment; but a new `SavingParameters` obj is dispatched in the queue by the `State`, as this happens in multiple places - basically any time we change something, even state probably.

**Comment by diegoasua** [comment]
> The problem with the duration is the following:
> At the moment we are setting a default panel of `SavingParameters`. The State dispatches them to the `StackSaver` via a new `SavingParameters()` object generated by the call to `convert_save_params()` function in `State`, line 599.
> This always dispatches to the `StackSaver` the standard duration, which has always been overwritten by the value in the `duration_queue` which has always been arrived directly from the `ExternalComm` process.
> Now we are changing it in the GUI before starting the experiment; but a new `SavingParameters` obj is dispatched in the `saving_parameter_queue` by the `State`, as this happens in multiple places - basically any time we change something, even state probably.

The problem is not in `State`. The problem lays on `streaming_save.py`:

https://github.com/portugueslab/sashimi/blob/df83b9974b78b766cba972f5e0df7387cb2744c6/sashimi/processes/streaming_save.py#L101

Duration gets overwritten when saving loop starts. This was intended for correcting for multiple experiments in the same session of the software. In triggered mode duration arrives after `Saver` has started. In no-trigger mode duration is set before.

**Comment by vigji** [comment]
> > The problem with the duration is the following:
> > At the moment we are setting a default panel of `SavingParameters`. The State dispatches them to the `StackSaver` via a new `SavingParameters()` object generated by the call to `convert_save_params()` function in `State`, line 599.
> > This always dispatches to the `StackSaver` the standard duration, which has always been overwritten by the value in the `duration_queue` which has always been arrived directly from the `ExternalComm` process.
> > Now we are changing it in the GUI before starting the experiment; but a new `SavingParameters` obj is dispatched in the `saving_parameter_queue` by the `State`, as this happens in multiple places - basically any time we change something, even state probably.
> 
> The problem is not in `State`. The problem lays on `streaming_save.py`:
> 
> https://github.com/portugueslab/sashimi/blob/df83b9974b78b766cba972f5e0df7387cb2744c6/sashimi/processes/streaming_save.py#L101
> 
> Duration gets overwritten when saving loop starts. This was intended for correcting for multiple experiments in the same session of the software. In triggered mode duration arrives after `Saver` has started. In no-trigger mode duration is set before.

Yes exactly, the overwriting happens between the 2 different try-except of the `receive_save_parameters`

**Comment by vigji** [comment]
Now, I have 2 solutions to offer:
1) We have a `self.n_volumes` attribute in the `State` object, that is updated when changing the parameter from the GUI; and around line 562 of the State, we always dispatch this to the `experiment_duration_queue` every time we send new parameters; even if we don't change it manually, it will always been overwritten by the duration sent by Stytra in the triggered experiment.  
**pros:** quick and not interfering with existing safe code; not adding new queues; 
**cons:** duration still dispatched through 2 entry points, and is been overwritten - although that is already the case).
2) We dispatch the duration from the external communication to the State, and then dispatch it together with the rest of the parameters; but given the complex timeline at the experiment starting, this might be dangerous. The `duration_queue` would then only end in the `State`, and not in the `StackSaver`.
**pros**: enforce single entri point for the duration;
**cons**: rerouting of a queue, potentially interfering with current sequence of events at experiment start

ATM I would personally be more inclined toward 1

**Comment by diegoasua** [comment]
Fixed

**Comment by diegoasua** [comment]
There are some glitches with `QWidget.show(0` and `QWidget.hide()`. If you go front and back with `Pause` and `Volume` modes the bar and saving status properly hide, but if you do it a second time they don't. Not sure if it is on our side or on `PyQt5` side. Anyhow, that is just UX and does not affect the normal functioning of the software. It can be addressed later.

**Comment by vigji** [comment]
> There are some glitches with `QWidget.show()` and `QWidget.hide()`. If you go front and back with `Pause` and `Volume` modes the bar and saving status properly hide, but if you do it a second time they don't. Not sure if it is on our side or on `PyQt5` side. Anyhow, that is just UX and does not affect the normal functioning of the software. It can be addressed later. I left it as enabled/disabled instead of show/hide.
> 
> So things to be done before merging this PR
> 
> 1. Make the `MockCamera` stop in `Paused` mode.
> 2. Properly run a full mock experiment in no-trigger mode, verify what was saved is actually what was intended.
> 
> Things that can be done here or in further PRs:
> 
> 1. @vilim 's tab idea. This is very cool and professional but might involve a lot of debugging. I would be in favor of opening a further PR about UX and aesthetics once this is merged, so that we can proceed with #92 and work on planar mode, which should be the number one priority right now.
> 2. Minor bugs with the toolbar and status bar, like the show/hide thing or sometimes inconsistent info shown in the status bar.
> 
> I would be in favour of wrapping up this with 1 and 2 and opening 3 and 4 in another PR. This has been and is being quite a painful PR and will sleep better knowing it is buried down.

Agree on all points! Have you fixed also the duration bug or we should go further with my solution?

**Comment by diegoasua** [comment]
The duration bug is fixed in`State` by modifying the function that curates parameters sent to `Saver`

**Comment by vigji** [comment]
We agreed to remove the duration parameter from the `SavingParameters` class

**Comment by diegoasua** [comment]
Okay guys this is ready. We'll tackle the start/stop camera issue in #92 and UX in a new PR. Ready for review @kkoetter if no issues green light for merge! 🥂

**Comment by kkoetter** [comment]
I tested it on my computer and saved data has the right dimensions in both modes.
 The `n_volumes `parameter is not fully removed from the `SavingParameters ` class. I think in the future if we implement a planar triggered mode `state.py ` lines 254-259 might cause problems. Else i have no concerns right now.

**Comment by vigji** [comment]
> I tested it on my computer and saved data has the right dimensions in both modes.
> The `n_volumes `parameter is not fully removed from the `SavingParameters ` class. I think in the future if we implement a planar triggered mode `state.py ` lines 254-259 might cause problems. Else i have no concerns right now.

This now has been addressed, together with the bar refreshing input. I will merge this!

**Pull request merged**
